File: crates/taflow-core/src/stream/schaff_trend_cycle.rs
use super::operator_states::*;
use crate::error::{TaError, TaResult};

use super::operator_states::*;
use super::*;
use std::collections::{HashMap, HashSet, VecDeque};

#[derive(Debug, Clone, Copy, PartialEq)]
/// Persistent Rust state or aligned output type for `SchaffTrendCycleValue`.
///
/// The state consumes chronological inputs causally, preserves warm-up
/// values, and exposes the current result through its public API.
pub struct SchaffTrendCycleValue {
    pub stc: f64,
    pub macd: f64,
    pub stoch: f64,
}

/// Stateful Schaff Trend Cycle (pandas-ta classic `momentum/stc.py`, theory:
/// Douglas Schaff). MACD line from two SMA-seeded EMAs, then two cascaded
/// stochastics with `round(..., 8)` smoothing at `factor`.
///
/// The `stc`/`stoch` series are fully defined from bar 0 (seeded `0` and
/// carried forward while the rolling windows are cold or non-positive); the
/// `macd` line is NaN until both EMAs are warm.
#[derive(Debug, Clone)]
/// Persistent Rust state or aligned output type for `SchaffTrendCycle`.
///
/// The state consumes chronological inputs causally, preserves warm-up
/// values, and exposes the current result through its public API.
pub struct SchaffTrendCycle {
    tclength: usize,
    fast: usize,
    slow: usize,
    pub(crate) factor: f64,
    pub(crate) fast_ema: ExponentialMovingAverage,
    pub(crate) slow_ema: ExponentialMovingAverage,
    pub(crate) xmacd_low: RollingExtremum,
    pub(crate) xmacd_high: RollingExtremum,
    pub(crate) pf_low: RollingExtremum,
    pub(crate) pf_high: RollingExtremum,
    pub(crate) stoch1: f64,
    pub(crate) pf: f64,
    pub(crate) stoch2: f64,
    pub(crate) pff: f64,
    pub(crate) value: Option<SchaffTrendCycleValue>,
}

impl SchaffTrendCycle {
    /// Computes or updates `new` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn new(tclength: usize, fast: usize, slow: usize, factor: f64) -> TaResult<Self> {
        validate_period(tclength)?;
        validate_period(fast)?;
        validate_period(slow)?;
        if !(factor > 0.0) {
            return Err(TaError::InvalidParameter {
                name: "factor",
                value: factor.to_string(),
                reason: "must be > 0",
            });
        }
        let (fast, slow) = if slow < fast {
            (slow, fast)
        } else {
            (fast, slow)
        };
        Ok(Self {
            tclength,
            fast,
            slow,
            factor,
            fast_ema: ExponentialMovingAverage::new(fast)?,
            slow_ema: ExponentialMovingAverage::new(slow)?,
            xmacd_low: RollingExtremum::new(tclength, true)?,
            xmacd_high: RollingExtremum::new(tclength, false)?,
            pf_low: RollingExtremum::new(tclength, true)?,
            pf_high: RollingExtremum::new(tclength, false)?,
            stoch1: 0.0,
            pf: 0.0,
            stoch2: 0.0,
            pff: 0.0,
            value: None,
        })
    }

    /// Computes or updates `append` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn append(&mut self, close: f64) -> SchaffTrendCycleValue {
        let fast = self.fast_ema.append(close);
        let slow = self.slow_ema.append(close);
        let macd = match (fast, slow) {
            (Some(fast), Some(slow)) => fast - slow,
            _ => f64::NAN,
        };

        let lowest = self.xmacd_low.append(macd).unwrap_or(f64::NAN);
        let highest = self.xmacd_high.append(macd).unwrap_or(f64::NAN);
        let range = non_zero(highest - lowest);
        if lowest > 0.0 {
            self.stoch1 = 100.0 * ((macd - lowest) / range);
        }
        self.pf = round8(self.pf + self.factor * (self.stoch1 - self.pf));

        let lowest_pf = self.pf_low.append(self.pf).unwrap_or(f64::NAN);
        let highest_pf = self.pf_high.append(self.pf).unwrap_or(f64::NAN);
        let range_pf = non_zero(highest_pf - lowest_pf);
        if range_pf > 0.0 {
            self.stoch2 = 100.0 * ((self.pf - lowest_pf) / range_pf);
        }
        self.pff = round8(self.pff + self.factor * (self.stoch2 - self.pff));

        let value = SchaffTrendCycleValue {
            stc: self.pff,
            macd,
            stoch: self.pf,
        };
        self.value = Some(value);
        value
    }

    /// Bulk kernel for the MACD chain: once both EMAs are warm, their scalar
    /// recurrences advance in locals inside one loop; the two cascaded
    /// stochastic stages (rolling extrema + smoothing) advance in place with
    /// the exact per-bar arithmetic. Bit-identical to per-bar [`Self::append`]
    /// in outputs and post-run streaming state.
    pub fn extend_slices_into(
        &mut self,
        close: &[f64],
        stc_out: &mut Vec<f64>,
        macd_out: &mut Vec<f64>,
        stoch_out: &mut Vec<f64>,
    ) {
        schaff_trend_cycle_bulk(self, close, stc_out, macd_out, stoch_out)
    }
}

impl SchaffTrendCycle {
    /// Computes or updates `value` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn value(&self) -> Option<SchaffTrendCycleValue> {
        self.value
    }

    /// Computes or updates `reset` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn reset(&mut self) {
        self.fast_ema.reset();
        self.slow_ema.reset();
        self.xmacd_low.reset();
        self.xmacd_high.reset();
        self.pf_low.reset();
        self.pf_high.reset();
        self.stoch1 = 0.0;
        self.pf = 0.0;
        self.stoch2 = 0.0;
        self.pff = 0.0;
        self.value = None;
    }
}

File: python/taflow/schaff_trend_cycle.py
"""Persistent Schaff Trend Cycle (pandas-ta classic alignment)."""

from typing import Any
import numpy as np
from ._native import SchaffTrendCycleOperator as _Native
from ._series import as_float64_series


class SchaffTrendCycle:
    """Persistent Schaff Trend Cycle (pandas-ta classic alignment).

    This public class owns a persistent native Rust state; Python performs container conversion only. `append`, `extend`, and `reset` are fluent, `value` exposes the latest result, and `compute` returns aligned history. Required input histories: `close`. Warm-up positions are represented by `NaN` in history."""

    def __init__(
        self,
        close: Any,
        tclength: int = 10,
        fast: int = 12,
        slow: int = 26,
        factor: float = 0.5,
    ) -> None:
        """Initialize this adapter and process the supplied input series.

        Parameters
        ----------
        close : object
            Close-price series or the current bar close.
        tclength : object
            Schaff cycle length.
        fast : object
            Fast smoothing period in bars.
        slow : object
            Slow smoothing period in bars.
        factor : object
            Trend multiplier.

        Returns
        -------
        None
            The constructor initializes the adapter and returns no value.
        """
        self._state = _Native(tclength, fast, slow, factor)
        self._length = 0
        self.extend(close)

    def append(self, close: float) -> "SchaffTrendCycle":
        """Append one observation or aligned bar to the native Rust state.

        Parameters
        ----------
        close : object
            Close-price series or the current bar close.

        Returns
        -------
        Self
            The updated adapter, native value, aligned output array, or execution node.
        """
        self._state.append(float(close))
        self._length += 1
        return self

    def extend(self, close: Any) -> "SchaffTrendCycle":
        """Append aligned input series to the native Rust state.

        Parameters
        ----------
        close : object
            Close-price series or the current bar close.

        Returns
        -------
        Self
            The updated adapter, native value, aligned output array, or execution node.
        """
        close_array = as_float64_series(close)
        self._state.extend(close_array)
        self._length += len(close_array)
        return self

    def compute(self) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
        """Return the aligned output history as a NumPy array.

        Returns
        -------
        numpy.ndarray or tuple of numpy.ndarray
            The updated adapter, native value, aligned output array, or execution node.
        """
        return self._state.compute()

    @property
    def value(self) -> tuple[float, float, float] | None:
        """Return the latest computed value, or None during warm-up.

        Returns
        -------
        float, tuple, or None
            The updated adapter, native value, aligned output array, or execution node.
        """
        return self._state.value

    def reset(self) -> "SchaffTrendCycle":
        """Execute the reset operation through the native Rust implementation.

        Returns
        -------
        Self
            The updated adapter, native value, aligned output array, or execution node.
        """
        self._state.reset()
        self._length = 0
        return self

    def __len__(self) -> int:
        """Return the number of processed closes."""
        return self._length
